fluent-plugin-events/lib/fluent/plugin/in_events.rb

```diff
@@ -43,90 +44,139 @@ def configure(conf)
       def start
         super
         initialize_resource_version
-        start_watcher_thread
-        start_configmap_flush_thread
+        start_monitor
       end
   
-      def close
-        @watcher.each &:finish
+      def stop
+        log.debug "Clean up before stopping completely"
+        update_config_map
+        @watch_stream.finish if @watch_stream
         super
       end
 
-      def initialize_resource_version
-        # get or create the config map
-        begin
-          @client.public_send("get_config_map", "fluentd-config-resource-version", "sumologic").tap do |resource|
-            log.info "Get config maps: #{resource}"
-            version = resource.data['resource-version']
-            @resource_version = version if version
-          end
-        rescue Kubeclient::ResourceNotFoundError
-          create_config_maps
+      # Listen for SIGTERM but do not trap
+      def prepend_handler(signal, &handler)
+        previous = Signal.trap(signal) do
+          previous = -> { raise SignalException, signal} unless previous.respond_to?(:call)
+          handler.call(previous)
         end
       end
 
-      def create_config_maps
-        @resource_version = "0"
-        resource = ::Kubeclient::Resource.new
-        resource.metadata = {
-          name: "fluentd-config-resource-version",
-          namespace: "sumologic"
-        }
-        resource.data = { "resource-version": "#{@resource_version}" }
-        @client.public_send("create_config_map", resource).tap do |maps|
-          log.info "Created config maps: #{maps}"
+      def start_monitor
+        log.info "Starting events collection"
+
+        last_recreated = Time.now.to_i
+        log.debug "last_recreated initialized to #{last_recreated}"
+
+        interrupted = false
+        prepend_handler("TERM") do |old|
+          interrupted = true
+          old.call
+        end
+
+        while !interrupted do
+          # Periodically restart watcher connection by checking if enough time has passed since 
+          # last time watcher thread was recreated or if the watcher thread has been stopped.
+          now = Time.now.to_i
+          watcher_exists = Thread.list.select {|thread| thread.object_id == @watcher_id && thread.alive?}.count > 0
+          if now - last_recreated >= @watch_interval_seconds || !watcher_exists
+            
+            log.debug "Recreating watcher thread. Use resource version from latest snapshot if watcher is running"
```

Comment:
Do we need to put the value of `@resource_version` in the log line, maybe after `pull_resource_version`?

Reply:
We log the resource version being used to start the watch stream in `start_watcher_thread`
```
log.debug "New thread to watch events #{@watcher_id} from resource version #{params[:resource_version]}"
```

Reply:
That works.